onnxruntime/core/framework/sparse_tensor.cc

```diff
@@ -3,31 +3,441 @@
+      buffer_size_(0),
+      values_(),
+      format_data_() {
+}
+
+SparseTensor::SparseTensor(SparseTensor&& o) noexcept : SparseTensor() {
+  *this = std::move(o);
+}
+
+SparseTensor& SparseTensor::operator=(SparseTensor&& o) noexcept {
+  ReleaseBuffer();
+  format_ = o.format_;
+  dense_shape_ = std::move(o.dense_shape_);
+  ml_data_type_ = o.ml_data_type_;
+  allocator_ = std::move(o.allocator_);
+  location_ = std::move(o.location_);
+  std::swap(p_data_, o.p_data_);
+  std::swap(buffer_size_, o.buffer_size_);
+  values_ = std::move(o.values_);
+  format_data_ = std::move(o.format_data_);
+  return *this;
+}
+
+SparseTensor::~SparseTensor() {
+  ReleaseBuffer();
+}
+
+Status SparseTensor::AllocateBuffer(int64_t buffer_size, size_t num_values) {
+  if (buffer_size > 0) {
+    SafeInt<size_t> b_size_t(buffer_size);
+    ORT_ENFORCE(b_size_t.Ref() > (num_values * ml_data_type_->Size()),
+                "Values size must be less than total buffer size");
+    auto data_ptr = IAllocator::MakeUniquePtr<void>(allocator_, b_size_t);
+    ORT_RETURN_IF_NOT(data_ptr != nullptr, "SparseTensor Allocation failed for size: ", buffer_size);
+    if (IsDataTypeString()) {
+      // We own the buffer, so we must properly construct strings. Neither of the Tensors
+      // we construct on top of the buffer own it. We are constructing empty strings, hopefully
+      // nothrow and no buffer allocation
+      utils::ConstructStrings(data_ptr.get(), static_cast<int64_t>(num_values));
+    }
+    p_data_ = data_ptr.release();
+  }
+  buffer_size_ = buffer_size;
+  return Status::OK();
+}
+
+void SparseTensor::ReleaseBuffer() {
+  if (allocator_ && p_data_ != nullptr) {
+    // if current tensor is responsible for deleting the buffer
+    // and it is a string tensor, need to explicitly call string(s)
+    // __dtor(s).
+    utils::ReleaseTensorBuffer(allocator_, IsDataTypeString(), p_data_, values_.Shape().Size());
+  }
+  p_data_ = nullptr;
+  buffer_size_ = 0;
+}
+
+void SparseTensor::CopyStrings(const Tensor& src, Tensor& dst) const {
+  auto src_span = src.DataAsSpan<std::string>();
+  auto* dst_iter = dst.MutableData<std::string>();
+  std::copy(src_span.cbegin(), src_span.cend(), dst_iter);
+}
+
+Status SparseTensor::CopyData(const IDataTransfer& data_transfer,
+                              const std::vector<std::reference_wrapper<const Tensor>>& src,
+                              const std::vector<std::reference_wrapper<Tensor>>& dst) {
+  ORT_RETURN_IF_NOT(src.size() == dst.size(), "Must have the same size");
+  for (size_t i = 0; i < src.size(); ++i) {
+    const Tensor& src_t = src[i];
+    Tensor& dst_t = dst[i];
+    if (src_t.IsDataTypeString()) {
+      CopyStrings(src_t, dst_t);
+    } else {
+      ORT_RETURN_IF_ERROR(data_transfer.CopyTensor(src_t, dst_t));
+    }
+  }
+  return Status::OK();
+}
+
+SparseTensor::CooView SparseTensor::AsCoo() const {
+  ORT_ENFORCE(Format() == SparseFormat::kCoo, "Must contain Coo format");
```

Comment:
Can you print the offending format as well? Same for many other places. Let's make our exceptions and error messages as informative and descriptive as we can.